File: Computer Programming/C++/Recursion and Dynamic Programming/Combination Sum III.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) 
    {
        vector<vector<int>> comb;
        vector<int> current;
        vector<int> number;

        for(int i=1;i<=9;i++)
        {
            number.push_back(i);
        }
        int index = 0;
        function(comb,current,number,k,n,index);   
        return comb; 
    }

    void function(vector<vector<int>>&comb, vector<int>&current,vector<int>&number,int k,int n,int index)
    {
        if(k==0 && n==0)
        {
            comb.push_back(current);
            return;
        }

        if(k<0 || n<0 || index > 9)
        {
            return;
        }

        function(comb,current,number,k,n,index+1);
        current.push_back(number[index]);
        function(comb,current,number,k-1,n-number[index],index+1);
        current.pop_back();
        return;

    }
};

File: Computer Programming/Python/Recursion/Combination Sum III.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        
        ans = []
        current = []

        self.function(ans,current,9,k,n)
        return ans

    def function(self,ans,current,index,k,target):
        if target == 0:
            if k == 0:
                ans.append(list(current))
                return
            else:
                return

        if target < 0:
            return 
        if k < 0:
            return
        if index < 0:
            return

        current.append(index)
        self.function(ans,current,index-1,k-1,target - index)
        current.pop()

        self.function(ans,current,index-1,k,target)

        return

        
